semantic/types.go

```diff
@@ -113,6 +115,53 @@ func (k Kind) ReturnType() Type {
 }
 func (k Kind) typ() {}
 
+// tvar represents a type variable for type inference
+type tvar struct {
+	name string
+}
+
+func (tv *tvar) String() string {
+	return tv.name
+}
+func (tv *tvar) Kind() Kind {
+	return TypeVar
+}
+func (tv *tvar) PropertyType(name string) Type {
+	panic(fmt.Errorf("cannot get property type of kind %s", tv.Kind()))
+}
+func (tv *tvar) Properties() map[string]Type {
+	panic(fmt.Errorf("cannot get properties type of kind %s", tv.Kind()))
+}
+func (tv *tvar) ElementType() Type {
+	panic(fmt.Errorf("cannot get element type of kind %s", tv.Kind()))
+}
+func (tv *tvar) Params() map[string]Type {
+	panic(fmt.Errorf("cannot get parameters from kind %s", tv.Kind()))
+}
+func (tv *tvar) PipeArgument() string {
+	panic(fmt.Errorf("cannot get pipe argument name from kind %s", tv.Kind()))
+}
+func (tv *tvar) ReturnType() Type {
+	panic(fmt.Errorf("cannot get return type of kind %s", tv.Kind()))
+}
+func (tv *tvar) typ() {}
+
+var typeVarCache = struct {
+	prefix string
+	suffix int
+}{
+	prefix: "t",
+	suffix: 0,
+}
+
+// NewTypeVar constructs a new type variable to be assigned to an expression
+// Type variable names progress as such - t0, t1, t2, t3, ...
+func NewTypeVar() Type {
+	name := typeVarCache.prefix + strconv.Itoa(typeVarCache.suffix)
+	typeVarCache.suffix++
```

Comment:
This is not memory safe. If multiple Annotation calls are running concurrently then this will race and possibly crash.

 Each Annotation phase should create its own typeVarCache, that way the type vars are all unique. We should not add locking here as there is no need for the vars to be global across different annotation phases.

Reply:
Agreed, will update.